### PR TITLE
Remove ObjectServerUser #3741 #4962 #5028

### LINKED ISSUE #3741: Migrate SyncUser to the ObjectStore
SyncUser class should use the ObjectStore SyncUser.

### PR DESCRIPTION
**Pull request by nhachicha**
- Removed `ObjectServerUser` to simplify the handling of `SyncUser` (reduce complexity)
- Fixed bugs in `logout` where it was possible to logout without checking if the associated Realm instances are still open.

fixes #3741 #4962 #5028

### CODE REVIEW

**Comment by Zhuinden** [comment]
> Fixed bugs in logout where it was possible to logout without checking if the associated Realm instances are still open.

I'm pretty sure @beeender said this is [hard to guarantee in a multi-process environment](https://github.com/realm/realm-java/issues/4962#issue-242580199).

**Comment by nhachicha** [comment]
@Zhuinden yeah, this is not intended for multi-process. We still don't support officially [multi-process](https://github.com/realm/realm-java/issues/1091).

**Comment by nhachicha** [comment]
@realm/java

File: realm/realm-library/src/syncIntegrationTest/java/io/realm/objectserver/SyncSessionTests.java [hunk 1]
```diff
@@ -106,4 +109,52 @@ public void uploadDownloadAllChanges() throws InterruptedException {
...
+                .createSyncConfigurationBuilder(user, Constants.SYNC_SERVER_URL_2)
+                .build();
+        Realm realm2 = Realm.getInstance(syncConfiguration2);
+
+        SyncSession session1 = SyncManager.getSession(syncConfiguration1);
+        SyncSession session2 = SyncManager.getSession(syncConfiguration2);
+
+        // make sure the `access_token` is acquired. otherwise we can still be
+        // in WAITING_FOR_ACCESS_TOKEN state
+        SystemClock.sleep(TimeUnit.SECONDS.toMillis(2));
```
Comment (beeender) [inline realm/realm-library/src/syncIntegrationTest/java/io/realm/objectserver/SyncSessionTests.java]: making a while loop to check the session state would be more efficient.

File: realm/realm-library/src/main/cpp/io_realm_SyncUser.cpp [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <jni.h>
+#include <string>
+
+#include "io_realm_SyncUser.h"
+
+#include "object-store/src/sync/sync_manager.hpp"
```
Comment (beeender) [inline realm/realm-library/src/main/cpp/io_realm_SyncUser.cpp]: -> `#include "sync/sync_manager.hpp"`
same for below.

File: realm/realm-library/src/main/cpp/io_realm_SyncUser.cpp [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+#include "object-store/src/sync/sync_session.hpp"
+
+#include "util.hpp"
+
+using namespace realm;
+
+static const char* ERR_COULD_NOT_ALLOCATE_MEMORY = "Could not allocate memory to return all sessions path.";
+
+JNIEXPORT jobjectArray JNICALL
+Java_io_realm_SyncUser_nativeAllSessionsPath(JNIEnv *env, jclass, jstring sync_user_identity) {
```
Comment (beeender) [inline realm/realm-library/src/main/cpp/io_realm_SyncUser.cpp]: format is wrong:
```
JNIEXPORT jobjectArray JNICALL Java_io_realm_SyncUser_nativeAllSessionsPath(JNIEnv* env, jclass,
                                                                            jstring sync_user_identity)
{
```

try to install a `clang-format` and use `./realm/realm-library/src/main/cpp/.clang-format` to format the code.

File: realm/config/findbugs/findbugs-filter.xml [hunk 4]
```diff
@@ -21,5 +21,10 @@
     <Match>
         <Class name="io.realm.PermissionOfferResponseRealmProxy"/>
     </Match>
+    <Match>
```
Comment (beeender) [inline realm/config/findbugs/findbugs-filter.xml]: can we just remove the local var `final Future<?> future = ` at `SyncUser.java:278`? maintaining a long list of findbugs filter is quite annoy.

Comment (bmeike) [inline realm/config/findbugs/findbugs-filter.xml]: I'm pretty sure that you can just use an annotation for this: @SuppressFBWarnings("DLS_DEAD_LOCAL_STORE"). That would be a lot better:

File: realm/realm-library/src/androidTestObjectServer/java/io/realm/SyncUserTests.java [hunk 5]
```diff
@@ -79,6 +99,49 @@ public void setUp() {
         SyncManager.reset();
     }
 
+    private static SyncUser createFakeUser(String id) {
+        final Token token = new Token("token_value", id, "path_value", Long.MAX_VALUE, null);
+        try {
+            return SYNC_USER_CONSTRUCTOR.newInstance(token, authUrl);
+        } catch (InstantiationException e) {
```
Comment (bmeike) [inline realm/realm-library/src/androidTestObjectServer/java/io/realm/SyncUserTests.java]: We are using Java 8 here.  You can say `catch InstantiationException | IllegalAccessException | ... `

File: realm/realm-library/src/syncIntegrationTest/java/io/realm/objectserver/SyncSessionTests.java (hunk 1 above)
Comment (bmeike) [inline realm/realm-library/src/syncIntegrationTest/java/io/realm/objectserver/SyncSessionTests.java]: Yeah... this is how flaky tests are made.  Is there any way to set a CountdownLatch to let the proceed?

**Comment by nhachicha** [comment]
- Aligned the implementation of the `SyncUser#logout` with other bindings, now `SyncUser#logout` will not throw an exception when remaining instances are not closed.

_Login again the same user will re-activate the logged out user. It will also revive inactive session(s) to resume uploading/downloading_ 

- The current behaviour of the ObjectStore is to remove the Realm of a logged out user (after app restart) this behaviour might be broken because of https://github.com/realm/realm-java/issues/5097 (_the current test is disabled_).

Even when #5097 is fixed, this behaviour is questionable for the following arguments https://realmio.slack.com/archives/C033Z2Y8T/p1502462340709245

- Changed the SyncStop policy session, to `SyncSessionStopPolicy:: AfterChangesUploaded`fixes #5028

**Comment by nhachicha** [comment]
Currently, tests are blocked by this issue https://github.com/realm/realm-java/issues/5097

File: CHANGELOG.md [hunk 6]
```diff
@@ -1,16 +1,22 @@
 ## 3.6.0 (YYYY-MM-DD)
 
+### Breaking Changes
+
+* [ObjectServer] `SyncUser.logout()` no longer throw an exception when associated Realms instances are not closed (#4962).
```
Comment (dalinaum) [inline CHANGELOG.md]: no longer throws?

File: CHANGELOG.md [hunk 7]
```diff
@@ -1,16 +1,22 @@
...
 * [ObjectServer] APIs of `UserStore` have been changed to support same user identity but different authentication server scenario.
-* [ObjectServer] Added `SyncUser.allSessions` to retrive the all valid sessions belonging to the user (#4783).
+* [ObjectServer] Added `SyncUser.allSessions` to retrieve the all valid sessions belonging to the user (#4783).
 * Added `Nullable` annotation to methods that may return `null` in order to improve Kotlin usability. This also introduced a dependency to `com.google.code.findbugs:jsr305`.
 * Added support for new data type `MutableRealmIntegers`. The new type behaves almost exactly as a reference to a Long (mutable nullable, etc) but supports `increment` and `decrement` methods, which implement a Conflict Free Replicated Data Type, whose value will converge even when changed across distributed devices with poor connections (#4266).
 
 ### Bug Fixes
 
 ### Internal
+* [ObjectServer] removed `ObjectServerUser` and it's inner classes, in a step to reduce `SyncUser` complexity (#3741).
```
Comment (dalinaum) [inline CHANGELOG.md]: its inner classes

File: CHANGELOG.md [hunk 8]
```diff
@@ -1,16 +1,22 @@
...
 
+### Breaking Changes
+
+* [ObjectServer] `SyncUser.logout()` no longer throw an exception when associated Realms instances are not closed (#4962).
+
 ### Enhancements
 
 * [ObjectServer] Added `SyncSession.uploadAllLocalChanges()`.
 * [ObjectServer] APIs of `UserStore` have been changed to support same user identity but different authentication server scenario.
-* [ObjectServer] Added `SyncUser.allSessions` to retrive the all valid sessions belonging to the user (#4783).
```
Comment (dalinaum) [inline CHANGELOG.md]: Ups, I had a mistake.

**Comment by nhachicha** [comment]
@realm/java

File: tools/sync_test_server/configuration.yml [hunk 9]
```diff
@@ -62,7 +62,7 @@ auth:
     ## The validity duration for Access Tokens. This should be a fairly small
     ## number, especially if you are concerned with revocations being applied
     ## quickly. This value is represented in seconds. Default: 1 minute.
-    # access_token: 60
+    access_token: 20
```
Comment (cmelchior) [inline tools/sync_test_server/configuration.yml]: You might want to add a comment here that changing this value can break a test and reference that test

File: realm/realm-library/src/main/cpp/CMakeLists.txt [hunk 10]
```diff
@@ -51,7 +51,7 @@ set(jni_headers_PATH /./${PROJECT_BINARY_DIR}/jni_include)
 if (build_SYNC)
     list(APPEND classes_LIST
         io.realm.ClientResetRequiredError io.realm.RealmFileUserStore
-        io.realm.SyncManager io.realm.SyncSession
+        io.realm.SyncManager io.realm.SyncSession io.realm.SyncUser
```
Comment (beeender) [inline realm/realm-library/src/main/cpp/CMakeLists.txt]: This change is not needed?

**Pull request merged**